---
title: "Alevin Index comparisons"
author: "Joshua Shapiro for CCDL"
output: html_notebook
---

This notebook contains analysis of some intial benchmark in runs of Alevin usign different index files, looking at mapping rates.

### Load Libraries
```{r setup}
library(ggplot2)
library(magrittr)
library(SingleCellExperiment)
library(tximport)
```

### File and directory setup
```{r}
data_dir <- file.path("data", "alevin-quant")
dir.create(data_dir, recursive = TRUE, showWarnings = FALSE)

quant_s3 <- "s3://nextflow-ccdl-results/scpca-benchmark/alevin-quant"

```




## Sync S3 files

```{r}
sync_call <- paste("aws s3 sync", quant_s3, data_dir)
system(sync_call, ignore.stdout = TRUE)
```

Get the sample list and make a data frame with sample info

```{r}
quant_dirs <- list.dirs(data_dir, full.names = FALSE, recursive = FALSE)

# split ids into components for later processing
quant_info <- data.frame (quant_dir = quant_dirs, info = quant_dirs) %>%
  tidyr::separate(info, sep = "[-]", 
                  into = c("sample", "index_type")) %>%
  tidyr::separate(index_type, 
                  into = c("index_content", "kmer", "sa"), 
                  extra = "drop") %>%
  dplyr::mutate(kmer = stringr::str_remove(kmer, "k"))
  
```

## Get Annotations from AnnotationHub

```{r}
hub = AnnotationHub::AnnotationHub(ask = FALSE)
# Ensembl v100 Homo Sapiens is AH79689
ensdb = hub[["AH79689"]]
ensg <- genes(ensdb)
```

Create vectors of mitochondiral genes and coding genes for later.

```{r}
# create the mitochondrial gene list
mito_genes <- ensg[seqnames(ensg) == 'MT']$gene_id

coding_genes <- ensg[ensg$gene_biotype == "protein_coding"]$gene_id
```



# Process Alevin quantifications

## tximport and make SCEs

```{r}
sces <- quant_dirs %>%
  purrr::map(
    ~ tximport(file.path(data_dir, .x, "alevin", "quants_mat.gz"),
               type = "alevin")) %>%
  purrr::map(
    ~ SingleCellExperiment(list(counts = .x$counts))) 

# add names
names(sces) <- quant_dirs
```


Calculate cell and feature QC statistics for each sample.

```{r}
sces <- sces %>% 
  purrr::map(
    ~ scater::addPerCellQC(
      .x,
      subsets = list(mito = mito_genes[mito_genes %in% rownames(.x)],
                     ncRNA = rownames(.x)[!rownames(.x) %in% coding_genes] )
    ) 
  ) %>%
  purrr::map(scater::addPerFeatureQC)
```

Combine all cell QC stats into a single data frame
```{r}
sce_cell_qc <- purrr::map_df(sces, 
                        ~ as.data.frame(colData(.x)) %>%
                          tibble::rownames_to_column(var = "cell_id"), 
                        .id = "quant_id") %>%
  dplyr::left_join(quant_info, by = c("quant_id" = "quant_dir"))
```


Combine all feature QC stats into a single data frame.
```{r}
sce_feature_qc <- purrr::map_df(sces, 
                        ~ as.data.frame(rowData(.x)) %>%
                          tibble::rownames_to_column(var = "gene_id"), 
                        .id = "quant_id") %>%
  dplyr::left_join(quant_info, by = c("quant_id" = "quant_dir"))
```

## Plotting the QC stats

Plot cell QC gene accumulation curves & mito fraction
```{r}
ggplot(sce_cell_qc %>% dplyr::filter(kmer == "31"), 
       aes(x = sum, y = detected, color = subsets_mito_percent)) +
  geom_point() +
  facet_grid(sample ~ paste(index_content, sa) ) +
  scale_color_viridis_c() +
  scale_x_log10() +
  scale_y_log10()
  
```

# Find features that are reliably detected

To start, will select features that are detected in > 5% of cells in at least one sample/index. 

```{r}
detected_features <- sce_feature_qc %>%
  dplyr::filter(detected >= 5.0) %>%
  dplyr::pull(gene_id) %>%
  unique()
# How many?
length(detected_features)
```

How many of these are not protein coding?

```{r}
detected_nc <- detected_features[!detected_features %in% coding_genes]

length(detected_nc)
```

Lets look at these features in a bit more depth, comparing the same sample with full transcriptome vs cDNA only.
```{r}
# select noncoding detected features
nc_feature_qc <- sce_feature_qc %>% 
  dplyr::filter(gene_id %in% detected_nc) 

# Get detected ncRNAs in 834-txome_k31_no_sa as a transcriptome 

nc_txome <-  nc_feature_qc %>%
  dplyr::filter(quant_id == "834-txome_k31_no_sa") %>%
  dplyr::arrange(desc(mean))

nc_cdna <-  nc_feature_qc %>%
  dplyr::filter(quant_id == "834-cdna_k31_no_sa")

# Join the tables
nc_feature_comparison <- nc_txome %>%
  dplyr::left_join(nc_cdna, 
                   by = "gene_id", 
                   suffix = c("_txome", "_cdna")) %>%
  dplyr::select("gene_id", "mean_txome", "mean_cdna", "detected_txome", "detected_cdna")

nc_feature_comparison
```

As expected, most of the features that are not listed as coding do not appear in the cdna only set!
Some do, and spot checking indicates that those are listed as pseudogenes, which are for whatever reason still included in the cDNA file.

Of those that do not, the top few genes are MALAT1 and mitchondrial rRNAs. 
MALAT1 is potentially interesting, and it does seem that we should probably keep it!

Lets make a vector of the noncoding cDNAs for future work:
```{r}
nc_cdna_genes <- sce_feature_qc %>%
  dplyr::filter(quant_id == "834-cdna_k31_no_sa",
                ! gene_id %in% coding_genes) %>%
  dplyr::pull(gene_id)
```


## Comparing coding and noncoding txomes

First, looking at the mean expression of each gene, comparing coding and noncoding: comparing within samples, coding to noncoding.

```{r}
sce_feature_filtered <- sce_feature_qc %>% 
  dplyr::filter(kmer == "31", sa == "no") %>%
  dplyr::mutate(class = dplyr::case_when(gene_id %in% coding_genes ~ "coding",
                                         gene_id %in% nc_cdna_genes ~ "noncoding cDNA",
                                          TRUE ~ "ncRNA"))

ggplot(sce_feature_filtered, 
       aes (x = mean, color = class)) +
  geom_density() + 
  scale_x_log10() + 
  scale_color_brewer(palette = "Dark2") +
  facet_grid(index_content ~ sample)
```

The noncoding cDNA genes seem to follow a very similar ditribution to the noncoding cDNA genes (mostly pseudogenes?), but I can't see any particular reason to exclude them. 
If anything, there are ncRNA genes tend to have somewhat higher mean expression than the noncoding cDNAs which are in the cDNA dataset.

Just to check, lets look at the correlation of coding genes within a sample:


```{r}
# pick a random set of cells to look at
cell_sample <- sample(colnames(sces[["834-cdna_k31_no_sa"]]), 100)

features <- rownames(sces[["834-cdna_k31_no_sa"]])
coding_features <- features[features %in% coding_genes]

compare_expression <- data.frame(
  cdna = as.vector(counts(sces[["834-cdna_k31_no_sa"]][coding_features, cell_sample])),
  txome = as.vector(counts(sces[["834-txome_k31_no_sa"]][coding_features, cell_sample]))
)

cor(compare_expression, method = "spearman")
```
```{r}
qqplot(log1p(compare_expression$cdna), log1p(compare_expression$txome))
```
Correlation is nearly perfect.

### Transcriptome Conclusion

I think there is little cost, and a not insignificant potential gain to including noncoding RNA transcripts in the mapping sets for this project.

## Comparing decoy sequences (TBD)




