---
title: "SCRNAseq quantification comparisons: Data Import"
author: "Joshua Shapiro for CCDL"
output: 
  html_notebook:
    toc: true
    toc_float: true
---

This notebook contains code to import and do preliminary processing single cell expression data, quantified by Alevin, Kallisto, and Cellranger. 
Alevin samples were quantified with and without decoy sequences. 
The imported data is stored as SingleCellExperiment objects, which are then saved as an `.rds` file for later analysis.

## Setup

### Load Libraries for import
```{r setup}
library(ggplot2)
library(magrittr)
library(SingleCellExperiment)
library(DropletUtils)
library(tximport)

# set seed
set.seed(2020)
```


### File and directory setup

Input file locations first, local and remote (S3)
```{r}
base_dir <- here::here()

data_dir <- file.path(base_dir, 'data', 'quants')
dir.create(data_dir, recursive = TRUE, showWarnings = FALSE)

alevin_data_dir <- file.path(data_dir, 'alevin')
dir.create(alevin_data_dir, recursive = TRUE, showWarnings = FALSE)
quant_s3_alevin <- 's3://nextflow-ccdl-results/scpca/alevin-quant'

kallisto_data_dir <- file.path(data_dir, 'kallisto')
dir.create(kallisto_data_dir, recursive = TRUE, showWarnings = FALSE)
quant_s3_kallisto <- 's3://nextflow-ccdl-results/scpca/kallisto-quant'

cellranger_data_dir <- file.path(data_dir, 'cellranger')
dir.create(cellranger_data_dir, recursive = TRUE, showWarnings = FALSE)
quant_s3_cellranger <- 's3://nextflow-ccdl-results/scpca/cellranger-quant'
```

Output files will be `.rds` files, stored in a subdirectory of this notebook's location.
```{r}
export_dir <- 'data'

alevin_rds <- file.path(export_dir, 'alevin_sces.rds')
kallisto_rds <- file.path(export_dir, 'kallisto_sces.rds')
cellranger_rds <- file.path(export_dir, 'cellranger_sces.rds')

```



### Sync S3 files

```{r}
samples = c('SCPCR000001', 'SCPCR000002')
# generate include statements for sample directories
includes = stringr::str_glue("--include \"*/{samples}*\"")

# Alevin quantification files
sync_call <- paste('aws s3 sync', quant_s3_alevin, alevin_data_dir,
                   includes)
system(sync_call, ignore.stdout = TRUE)

# Kallisto quantification files
# exclude bus files, which are large
sync_call <- paste('aws s3 sync', quant_s3_kallisto, kallisto_data_dir,
                   includes, '--exclude "*/bus/*"')
system(sync_call, ignore.stdout = TRUE)

# Cellranger quantification files
# exclude intermediate and bam files
sync_call <- paste('aws s3 sync', quant_s3_cellranger, cellranger_data_dir,
                   includes, 
                   '--exclude "*/SC_RNA_COUNTER_CS/*"',
                   '--exclude "*.bam"', '--exclude "*.bam.bai"')
system(sync_call, ignore.stdout = TRUE)
```


### Generate sample info data frame

```{r}
quant_dirs <- list.dirs(c(alevin_data_dir, kallisto_data_dir, cellranger_data_dir),
                        full.names = TRUE, recursive = FALSE) %>%
  stringr::str_split("/") %>%
  purrr::map_df( ~ data.frame(tool = .x[3], quant_dir = .x[4], info = .x[4]))

# split ids into components for later processing
quant_info <- quant_dirs %>%
  tidyr::separate(info, sep = "[-]",
                  into = c("sample", "index_type")) %>%
  tidyr::separate(index_type, 
                  into = c("index_content", "kmer", "decoy"), 
                  extra = "drop") %>%
  # no decoy samples get NA here, convert to text.
  dplyr::mutate(decoy = tidyr::replace_na(decoy, "no"))
  
```

## Process quantification files

### Alevin
Use tximport to make SCEs

```{r}
alevin_quant_ids <- quant_info %>%
  dplyr::filter(tool == "alevin") %>%
  dplyr::pull(quant_dir)

alevin_sces <- alevin_quant_ids  %>%
  purrr::map(
    ~ tximport(file.path(alevin_data_dir, .x, "alevin", "quants_mat.gz"),
               type = "alevin")) %>%
  purrr::map(
    ~ SingleCellExperiment(list(counts = .x$counts))) 

# add names
names(alevin_sces) <- alevin_quant_ids
```

### Kallisto

Get file basenames (no extension) for import
```{r}
kallisto_quant_ids <- quant_info %>%
  dplyr::filter(tool == "kallisto") %>%
  dplyr::pull(quant_dir)

base_file <- file.path(kallisto_data_dir, kallisto_quant_ids, "counts", "gene_count")
```

#### Function for reading Kallisto files

Takes a base file name and reads the the `.mtx` file and associated `.barcodes.txt` and `.genes.txt` files.  
Optionally filters cells/barcodes based on the knee plot using `DropletUtils::barcodeRanks()`.
If filtering is not performed, these objects are quite large, as kallisto seems to do no filtering for empty cells.

```{r}
read_kallisto_counts <- function(base, cellfilter = c("knee", "none")){
  cellfilter <- match.arg(cellfilter)
  counts <- Matrix::readMM(paste0(base,".mtx"))%>%
    t() %>% # transpose to gene x cell orientation
    as("dgCMatrix") # compress sparse matrix
  dimnames(counts) <- list(readLines(paste0(base,".", "genes.txt")),
                           readLines(paste0(base,".barcodes.txt")))
  if(cellfilter == "knee"){
    # calculate ranks and find knee
    rank_df <- DropletUtils::barcodeRanks(counts)
    cells <- which(rank_df$total >= metadata(rank_df)$knee)
    counts <- counts[, cells]
  }
  return(counts)
}
```

Make a list of kallisto runs as SCEs
```{r}
kallisto_sces <- base_file %>%
  purrr::map(read_kallisto_counts) %>%
  purrr::map(~ SingleCellExperiment(list(counts = .x))) 

# add names
names(kallisto_sces) <- kallisto_quant_ids
```

### Cellranger

Use `DropletUtils::read10xCounts` to read cellranger counts as SCEs

```{r}
cellranger_quant_ids <- quant_info %>%
  dplyr::filter(tool == "cellranger") %>%
  dplyr::pull(quant_dir)

cellranger_sces <- cellranger_quant_ids  %>%
  purrr::map(
    ~ read10xCounts(file.path(cellranger_data_dir, .x,
                              "outs", "filtered_feature_bc_matrix.h5"),
                    col.names = TRUE)
  ) %>% 
  purrr::map(
    # get the portion before the "-" which includes the barcode
    function(x) {
      colnames(x) <- stringr::str_extract(colnames(x), "^([^-]+)")
      return(x)
    }
  )

# add names
names(cellranger_sces) <- cellranger_quant_ids
```

## Export SingleCellExperiment objects

