---
title: "MiQC Exploration for ScPCA"
author: "Ally Hawkins for CCDL"
output: 
  html_notebook:
    toc: true
    toc_float: true
---

In this notebook, we are interested in exploring the use of [miQC](https://github.com/greenelab/miQC) as an additional approach to filtering cells to remove any remaining cells that may not be viable or have low sequencing information. 

Here, I am particularly comparing the use of miQC to using hard thresholds, in particular removing cells with greater than 20% mitochondrial content, less than 500 genes detected per cell, and less than 1000 total counts per cell. 

For the most part, I followed the instructions in the [miQC vignette](https://github.com/greenelab/miQC/blob/main/vignettes/miQC.Rmd) for plotting and filtering. 

I'm choosing to compare 2 single-cell samples (SCPCR000126, SCPCR000127) and 2 single-nuclei samples (SCPCR000118, SCPCR000119). We are particularly interested in how miQC will handle single-nuclei samples, as they should not contain many reads corresponding to mitochondrial genes. 

## Setup

```{r}
library(magrittr)
library(ggplot2)
library(SingleCellExperiment)
library(scpcaTools)
library(gridExtra)
```


```{r}
# file paths to emptyDrops filtered sces from Alevin-fry 
base_dir <- here::here()
cr_like_results_dir <- file.path(base_dir, "data", "results")

cr_like_sce_file <- file.path(cr_like_results_dir, "alevin-fry-cr-like-em-emptydrops-200-sces.rds")

cr_like_sce <- readr::read_rds(cr_like_sce_file)
```



```{r}
# read in mito gene list
sample_info_dir <- file.path(base_dir, "sample-info")
mito_file <- file.path(sample_info_dir, "Homo_sapiens.GRCh38.103.mitogenes.txt")
mito_genes <- readr::read_tsv(mito_file, col_names = "gene_id")
mito_genes <- mito_genes %>%
  dplyr::pull(gene_id) %>%
  unique()
```


## Plot Metrics 

Before we can do any modeling and perform any filtering, let's look at the distribution of mitochondrial fraction per cell and unique genes detected per cell. 
The assumptions of miQC rely on having a distribution of mitochondrial reads and unique genes found. 
It uses a joint model to look at the proportion of reads mapping to mitochondrial DNA and the number of detected genes and determines the probability of a cell being compromised based on this proportion. 

To start we have to calculate the per cell statistics using `addPerCellQC()`. 
```{r}
# add per cell qc 
cr_like_sce <- cr_like_sce %>%
  purrr::map(scater::addPerCellQC, subsets = list(mito = mito_genes))
```

```{r}
plot_miQC_metrics <- function(sce, sample_name){
  p <- miQC::plotMetrics(sce) +
    ggtitle(sample_name)
  print(p)
}

purrr::iwalk(cr_like_sce, plot_miQC_metrics)
```
Already you can see that the plot shows a distribution of mitochondrial reads from 0-100% for the two single cell samples, but not for the single-nuclei samples. 
This could affect the ability of the model to distinguish compromised cells. 

## Plot Model 

Now we can actually look at the miQC model and what cells will be considered compromised and filtered out. 
Here, I'm plotting the probabilities that each cell is compromised as calculated by miQC, which cells would be filtered out using a given cutoff, and then comparing it to if we were to filter our cells using a pre determind threshold. 

```{r}
# function to test miQC modeling with different parameters on sce objects and show plots
model_sce <- function(sce, title, model_type = "linear", posterior_cutoff = 0.75){
  
  # plot distributions of total counts/cell, genes detected/cell and mito content
  coldata_df <- data.frame(colData(sce))
  manual_filter_plot <- ggplot(coldata_df, aes(x = sum, y = detected, color = subsets_mito_percent)) + 
    geom_point(alpha = 0.5) +
    scale_color_viridis_c() + 
    labs(x = "Total Count",
         y = "Number of Genes Expressed",
         color = "Mitochondrial\nFraction") + 
    geom_vline(xintercept = 500) +
    geom_hline(yintercept = 1000) +
    ggtitle(title) +
    theme_classic()
  
  # create model
  sce_model <- miQC::mixtureModel(sce, model_type)
  
  # plot cells colored by probability of cell being compromised or not
  model_plot <- miQC::plotModel(sce, sce_model) +
    theme_classic()
  
  # plot cells that would be filtered using model vs cells that would be filtered using manual QC cutoffs
  miQC_filter_plot <- miQC::plotFiltering(sce, sce_model, posterior_cutoff) + 
    geom_hline(yintercept = 20) +
    geom_vline(xintercept = 500) + 
    theme_classic()
  
  # arrange plots
  grid.arrange(manual_filter_plot, model_plot, miQC_filter_plot, nrow = 2)
}
```


```{r fig.height=5, fig.width=10}
purrr::iwalk(cr_like_sce, model_sce)
```
Using the suggested parameters from miQC we see that it does a fairly good job of keeping cells with low mito content and high number of genes detected for the single-cell samples, except for a few cases in 126 and 127 where cells with low mito and high unique genes are being thrown out.
For the single-nuclei samples however, it looks like the model is throwing away a lot more cells than we may want to probably due to the low range in mitochondrial percentages to begin with. 
Let's see if we alter some of the parameters used, like the cutoff for filtering, or the model used to compute the probability of compromised cells, and if that impacts the output. 

### Testing Posterior Probability Cutoffs

miQC will remove any cells that have a high probability of being compromised, by default these are cells with higher than 0.75 posterior probability. 
We can increase that threshold, which means that more cells will be kept and see if we are able to recover some of the cells that were being thrown out previously that we think should have been included (ie. cells with low mito that are being thrown out). 

```{r}
# change the posterior cutoff for miQC to keep throw away cells with > 0.9 probability of being compromised
purrr::iwalk(cr_like_sce, model_sce, posterior_cutoff = 0.9)
```
It doesn't appear that increasing the probability threshold rescues very many more cells with low mito content in SCPCR000126 and SCPCR000127 that are being filtered out. 

### Testing different mixture models 

In addition to altering the threshold used for filtering, we can also test the use of different mixture models to calculate the posterior probability of a cell being compromised. 
We will test use of both the spline and polynomial modes rather than the default linear model. 
The polynomial model uses a single polynomial to model the entire datset, while spline uses a piecewise continuous function of many polynomials to model the data set. 

```{r}
# test spline mixture model rather than linear 
purrr::iwalk(cr_like_sce, model_sce, model_type = "spline")
```

Use of the spline model definitely doesn't seem ideal with one sample, SCPCR000126, showing the reverse of what we would expect and only keeping cells with low mito content. 
For the other samples we see a similar trend that we had observed in the linear model. 

```{r}
# test polynomial mixture model rather than linear
purrr::iwalk(cr_like_sce, model_sce, model_type = "polynomial")
```

For the linear model, we still see that cells with low mito are getting excluded from SCPCR000118 and SCPCR000119, but now see that same trend in SCPCR000220 and SCPCR000221 as well. 
It appears to model the cells with very high number of unique genes and low mito content as cells to exclude when we would rather keep those. 

## Filter Cells

Now we can compare filtering using miQC vs. using the pre-determined cutoffs to see how the different filtering effects the overall number of cells that are removed. 
We can also look at the distribution of each of the metrics used for filtering, genes detected/cell, and mito content/cell to see how filtering affects the population of cells. 
Here we are using a threshold of 500 genes detected per cell and 10% mitochondrial reads.

```{r}
# function to filter sce using miQC 
miQC_filter <- function(sce, model_type = "linear", posterior_cutoff = 0.75){
  sce_model <- miQC::mixtureModel(sce, model_type)
  filtered_sce <- miQC::filterCells(sce, sce_model)
}

# filter using manual 
manual_filter <- function(sce, 
                          mito_cutoff = 20, 
                          genes_detected_cutoff = 500) {
  
  # get vector of cells to keep 
  cells_keep <- sce$detected > genes_detected_cutoff & 
    sce$subsets_mito_percent < mito_cutoff
  
  # filter sce
  filtered_sce <- sce[,cells_keep]
}

manual_filtered_sce <- purrr::map(cr_like_sce, manual_filter)
miQC_filtered_sce <- purrr::map(cr_like_sce, miQC_filter)
manual_filtered_sce_10 <- purrr::map(cr_like_sce, manual_filter, mito_cutoff = 10)
manual_filtered_sce_mito_only <- purrr::map(cr_like_sce, manual_filter, genes_detected_cutoff = 0)
```


```{r}
# function to grab number of cells from sce
get_num_cells <- function(sce){
  num_cells <- dim(sce)[2]
}

manual_cell_num <- manual_filtered_sce %>%
  purrr::map(get_num_cells) %>%
  as.data.frame()
rownames(manual_cell_num) <- c("manual_mito_20")

manual_cell_num_10 <- manual_filtered_sce_10 %>%
  purrr::map(get_num_cells) %>%
  as.data.frame()
rownames(manual_cell_num_10) <- c("manual_mito_10")

manual_cell_num_mito_only <- manual_filtered_sce_mito_only %>%
  purrr::map(get_num_cells) %>%
  as.data.frame()
rownames(manual_cell_num_mito_only) <- c("manual_mito_only")

miQC_cell_num <- miQC_filtered_sce %>%
  purrr::map(get_num_cells) %>%
  as.data.frame()
rownames(miQC_cell_num) <- c("miQC")

pre_filter_cell_num <- cr_like_sce %>%
  purrr::map(get_num_cells) %>%
  as.data.frame()
rownames(pre_filter_cell_num) <- c("pre_filtering")

# create dataframe for plotting with number of cells per sce 
combined_cell_num_df <- dplyr::bind_rows(manual_cell_num,
                                         miQC_cell_num,
                                         manual_cell_num_10,
                                         manual_cell_num_mito_only,
                                         pre_filter_cell_num) %>%
  tibble::rownames_to_column("filtering_method") %>%
  tidyr::pivot_longer(cols = starts_with("SCPCR"),
                      names_to = "sample", 
                      values_to = "number_cells")
```

```{r}
ggplot(combined_cell_num_df, aes(x = sample, y = number_cells, fill = filtering_method)) + 
  geom_col(position = "dodge") +
  theme(axis.text.x = element_text(angle = 90))
```


It looks like manual filtering (with these specific cutoff combinations) are removing more cells than miQC for single-cell, but more cells are removed in single-nuclei samples with either miQC than manual. 


## Conclusions

1. Using the default parameters for miQC with the linear mixture model gives the most consistent and reliable results where the majority of cells with low mito and high unique genes are classified as cells to keep, there are still some caveats.
2. Cells from single-nuclei samples do not show as high a range of mito content and therefore the mixture model suggests a cutoff that leads to cells with a lower mito content than in single-cell being classified as compromised. This however is slightly expected considering single-nuclei samples should have little to no mito reads.
3. From these results we decided that we would not actually perform any further filtering ourselves, but rather apply miQC to each sample using the linear mixture model to obtain the posterior probability of a cell being compromised. Then we would create a ccdl_suggests column to identify which cells we would suggest filtering.

Suggested next steps include comparing PCA or UMAP results using the different filtering thresholds to identify if any clusters corresponding to "bad cells" remain. 
Additionally, we will need to identify if we want to add any additional criteria besides the posterior probability computed by miQC to be included as a "keep" cell in the `ccdl_suggests` column that will be appended to the `colData` for every filtered sce object. 

## Session Info
```{r}
sessioninfo::session_info()
```

