---
title: "Spatial Transcriptomics Benchmarking"
author: "Ally Hawkins for CCDL"
output: 
  html_notebook:
    toc: true
    toc_float: true
---

In this notebook we are comparing the use of Alevin-fry and Spaceranger for quantifying spatial transcriptomics libraries. 
Two spatial transcriptomic libraries were quantified using Alevin-fry and Spaceranger and the results were combined following the [Alevin-fry tutorial](https://combine-lab.github.io/alevin-fry-tutorials/2021/af-spatial/). 
We are comparing that to if we were to not integrate the Spaceranger data with Alevin-fry and only use Spaceranger. 
For simplicity we will only look at one library, SCPCR000372.

## Set Up

```{r}
library(magrittr)
library(ggplot2)
library(SingleCellExperiment)
library(SpatialExperiment)
library(ggupset)
library(gridExtra)
```

```{r}
# load in benchmarking functions that will be used for copying data and generating sample tables
function_path <- file.path(".." ,"benchmarking-functions", "R")
file.path(function_path, list.files(function_path, pattern = "*.R$")) %>%
  purrr::walk(source)
```


```{r}
# set up file paths 
base_dir <- here::here()

# output folder to store alevin-fry and cellranger quants from S3
data_dir <- file.path(base_dir, "data", "spatial") 
s3_out <- file.path(data_dir, "data", "quants")

# results directory 
results_dir <- file.path(data_dir, "results")

# sample name
sample_id <- c("SCPCR000372")

# create directories
if(!dir.exists(data_dir)){
  dir.create(data_dir, recursive = TRUE)
}

if(!dir.exists(results_dir)){
  dir.create(results_dir, recursive = TRUE)
}
```


```{r}
mito_file <- file.path(base_dir, "sample-info", "Homo_sapiens.GRCh38.103.mitogenes.txt")
  
# read in mito genes 
mito_genes <- readr::read_tsv(mito_file, col_names = "gene_id")
mito_genes <- mito_genes %>%
  dplyr::pull(gene_id) %>%
  unique()
```


## Alevin-fry + Spaceranger versus Spaceranger Only

Now let's take a look at comparing the two methods of using Alevin-fry + Spaceranger to only Spaceranger for quantification. 
To do this, we will read in the Alevin-fry + Spaceranger combined and Spaceranger only `SpatialExperiment` objects separately and then merge them into one list before grabbing the per cell and per gene quality metrics. 

### Create Spatial Experiments 

```{r}
# get path to fry output directory 
fry_dir <- file.path(s3_out, "alevin-fry-knee", sample_id)
fry_dir <- paste0(fry_dir, "-spliced_intron_txome_k31-salign-cr-like-em-knee")

# paths to spatial folders 
cellranger_folder <- paste0(sample_id, "-cdna-spatial")
spatial_dir <- file.path(s3_out, "cellranger", cellranger_folder, "outs", "spatial")
```


```{r}
# read in combined fry and spaceranger spe 
fry_spe <- create_fry_spaceranger_spe(fry_dir, 
                                       spatial_dir, 
                                       sample_id)
```

```{r}
# spaceranger output paths
spaceranger_out <- file.path(s3_out, "cellranger", cellranger_folder, "outs")

# read in spaceranger output directly using read10XVisium
spaceranger_spe <- create_spaceranger_spe(spaceranger_out, sample_id)
```

### Per Spot QC Metrics

Now that we have read in the data and created our two `SpatialExperiment` objects, we can go ahead and combine them into one list and then calculate the per spot QC metrics using `scuttle::addPerCellQCMetrics().`

```{r}
# create one list with both spe's together
all_spe_list <- list(fry_spe, spaceranger_spe)
names(all_spe_list) <- c("alevin-fry", "spaceranger")

# calculate per cell QC and output to a combined data frame with plotting 
all_spe_list <- all_spe_list %>%
    purrr::map(
      ~ scuttle::addPerCellQCMetrics(.x, 
                                     subsets = list(mito = mito_genes[mito_genes %in% rownames(.x)]))) %>%
  purrr::map(scuttle::addPerFeatureQCMetrics)
```

After adding in the per spot QC metrics to both of the spe's, we want to extract the `colData` from each spe and create a data frame that we can use for plotting. 
We will also need some information about each sample and how it was run, so we will create a sample metadata table, `sample_info_df` that will then be merged with the `colData`. 

```{r}
# create sample info dataframe to be joined with per spot dataframe later
sample_info_df <- quant_info_table(data_dir= s3_out, 
                 tools = c("cellranger", "alevin-fry-knee"),
                 samples = sample_id) %>%
  # convert cellranger to spaceranger 
  dplyr::mutate(tool = ifelse(tool == "cellranger", "spaceranger", tool))
```

When we convert the `colData` to a data frame we use the custom function, `spatial_coldata_to_df()` to do so and apply it to each spe in our list. 
```{r}
# join coldata dataframe with sample info
coldata_df <- purrr::map_df(all_spe_list, spatial_coldata_to_df, .id = "tool") %>%
  # remove extra -1 from spaceranger barcodes
  dplyr::mutate(spot_id = gsub("-1", "", spot_id)) %>%
  dplyr::left_join(sample_info_df,
                   by = c("tool"))
```

Now we only want to filter our data frame to contain spots that are shared between both tools and those that are found to be overlapping with the tissue. 

```{r}
# identify shared spots only 
spot_counts <- coldata_df %>%  
  dplyr::count(spot_id, sample)

common_spots <- spot_counts %>%
  dplyr::filter(n == 2) %>%
  dplyr::pull(spot_id)

coldata_df_common <- coldata_df %>%
  dplyr::filter(spot_id %in% common_spots,
                # only include spots that overlap with tissue
                in_tissue == 1)
```

We will also need to filter the spe's directly based on spots that are present in the tissue, so we create a small function to do this and then apply it to both spe's in the list. 

```{r}
# we will also want to filter the spe's directly 
filter_spe <- function(spe){
  spe <- spe[, spatialData(spe)$in_tissue == 1]
}

all_spe_filter <- all_spe_list %>%
  purrr::map(filter_spe)
```

When we look at our results, we will also want to visualize them so we will make a custom function to plot the results.
Because of the bug that is present in `SpatialExperiment::read10XVisium()` function, we will make two separate functions, one for spaceranger only spes and one for alevin-fry + spaceranger spes.

```{r}
# custom function for plotting spe results and coloring by column of colData of choice
plot_fry_spe <- function(spe, sample, column){
  # plot spots only 
  p1 <- ggspavis::plotSpots(spe, 
                            x_coord = "pxl_col_in_fullres", 
                            y_coord = "pxl_row_in_fullres", 
                            annotate = column) +
    scale_color_viridis_c()
  
  # plot with tissue underneath
  p2 <- ggspavis::plotVisium(spe, 
                                   x_coord = "pxl_col_in_fullres", 
                                   y_coord = "pxl_row_in_fullres", 
                                   fill = column) +
    scale_fill_viridis_c()
  
  # arrange plots and add sample name as title 
  grid.arrange(p1, p2, nrow = 1, top = grid::textGrob(sample))
}


plot_spaceranger_spe <- function(spe, sample, column){
  # plot spots only 
  p1 <- ggspavis::plotSpots(spe, 
                            # switch the order of row and col to flip the image 
                            x_coord = "pxl_row_in_fullres", 
                            y_coord = "pxl_col_in_fullres", 
                            annotate = column) +
    scale_color_viridis_c() 
  
  # plot with tissue underneath
  p2 <- ggspavis::plotVisium(spe, 
                                   x_coord = "pxl_row_in_fullres", 
                                   y_coord = "pxl_col_in_fullres", 
                                   fill = column) +
    scale_fill_viridis_c()
  
  # arrange plots and add sample name as title 
  grid.arrange(p1, p2, nrow = 1, top = grid::textGrob(sample))
  
}
```

First we will look at the per cell metrics: mitochondrial reads per cell, total UMI per cell, and total genes detected per cell. 

```{r}
# % mitochondrial reads/ spot 
ggplot(coldata_df_common, aes(x = tool, y = subsets_mito_percent, fill = tool)) + 
  geom_boxplot() +
  theme_classic() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  ylab("Mito Percent") + 
  xlab("")
```
```{r}
plot_fry_spe(all_spe_filter$`alevin-fry`, "alevin-fry", "subsets_mito_percent")
```
```{r}
plot_spaceranger_spe(all_spe_filter$spaceranger, "spaceranger", "subsets_mito_percent")
```
Overall it looks like mitochondrial content is low and fairly similar across both tools. 


```{r}
# total UMI/ spot 
ggplot(coldata_df_common, aes(x = sum, color = tool)) + 
  geom_density() + 
  theme_classic() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  ylab("UMI/spot") + 
  xlab("")
```
```{r}
plot_fry_spe(all_spe_filter$`alevin-fry`, "alevin-fry", "sum")
```

```{r}
plot_spaceranger_spe(all_spe_filter$spaceranger, "spaceranger", "sum")
```

```{r}
# total genes/ spot 
ggplot(coldata_df_common, aes(x = detected, color = tool)) + 
  geom_density() + 
  theme_classic() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  ylab("Genes detected/spot") + 
  xlab("") 
```
```{r}
plot_fry_spe(all_spe_filter$`alevin-fry`, "alevin-fry", "detected")
```

```{r}
plot_spaceranger_spe(all_spe_filter$spaceranger, "spaceranger", "detected")
```
Generally it looks like both tools are fairly similar in the number of UMI/cell and genes detected/cell.

### Per Gene QC Metrics

Let's also look at the correlation of mean gene expression across shared genes. 
We will first need to calculate the per feature QC on the filtered spes after removing spots not present in the tissue and then grab the `rowData` and combine into a data frame used for plotting. 

```{r}
all_spe_filter <- all_spe_filter %>%
  purrr::map(scuttle::addPerFeatureQCMetrics)
```

```{r}
# grab rowdata and combine with sample info
rowdata_df <- purrr::map_df(all_spe_filter, spatial_rowdata_to_df, .id = "tool") %>%
  dplyr::left_join(sample_info_df,
                   by = c("tool"))
```

We then want to filter out any lowly detected genes, (detected < 5.0) and restrict our analysis to those genes that are found in both tools. 

```{r}
gene_counts <- rowdata_df %>% 
  # remove genes that have a low frequency of being detected
  dplyr::filter(detected >= 5.0) %>%
  dplyr::count(gene_id, sample)

# restrict to only common genes 
common_genes <- gene_counts %>%
  dplyr::filter(n == 2) %>%
  dplyr::pull(gene_id)

rowdata_df_common <- rowdata_df %>%
  dplyr::filter(
    (gene_id %in% common_genes) 
  )
```

```{r}
# create a table to calculate correlation between mean gene expression
rowdata_cor <- rowdata_df_common %>%
  dplyr::select(tool, gene_id, sample, mean) %>%
  # spread the mean expression stats to one column per caller
  tidyr::pivot_wider(id_cols = c(gene_id, sample),
                     names_from = c("tool"),
                     values_from = mean) %>%
  # drop rows with NA values to ease correlation calculation
  tidyr::drop_na()
```

```{r}
# look at correlation between the two tools
rowdata_cor %>% 
  dplyr::group_by(sample) %>%
  dplyr::summarize(
    alevin_fry_knee_spaceranger_cor = cor(`spaceranger`, `alevin-fry`, method = "spearman")
  )
```

```{r}
# mean gene expression across shared genes 
ggplot(rowdata_cor, aes(x = `spaceranger`, y = `alevin-fry`)) +
  geom_point(size = 0.5, alpha = 0.1) + 
  scale_x_log10() + 
  scale_y_log10() + 
  labs(x = "Spaceranger mean gene expression", y = "Alevin Fry Mean gene expression") + 
  theme_classic()
```
Correlation appears to be quite high between mean gene expression in Spaceranger and Alevin-fry, however, we do see that there is a subset of genes that have higher gene expression in Spaceranger than in Alevin-fry and are slightly off the diagonal. 

### Shared genes across tools

The final thing we will look at is the overlap of genes detected in each tool. 
First we need to filter the `SpatialExperiment` objects to only have spots that are found in both tools and then we filter to include only genes that are found in both indices before looking at the overlap. 

```{r}
# make a function to filter sces by subset of spots and re-calculate feature stats
filter_spe <- function(spe, spots){
  # remove "-1" at end of barcode for spaceranger spes
  colnames(spe) <- gsub("-1", "", colnames(spe))
  cells_to_keep <- colnames(spe) %in% spots
  rowData(spe) <- NULL
  spe[, cells_to_keep] %>%
    scuttle::addPerFeatureQCMetrics()
}
```

```{r}
# filter all spes to only have common spots 
spe_list_common <- all_spe_filter %>%
  purrr::map(filter_spe, spots = common_spots)

# grab rowdata from filtered sces 
rowdata_df_filtered <- purrr::map_df(spe_list_common, spatial_rowdata_to_df, .id = "tool") %>%
  dplyr::left_join(sample_info_df,
                   by = c("tool"))
```


```{r}
# get genes common in all tools
common_genes <- rowdata_df_filtered %>%
  dplyr::select(gene_id, tool) %>%
  dplyr::distinct() %>%
  dplyr::group_by(gene_id) %>%
  dplyr::tally() %>%
  # filter for genes found in both spaceranger and alevin-fry
  dplyr::filter(n == 2) %>%
  dplyr::pull(gene_id)

# filter rowdata_df to only include genes found in all tools and genes with mean > 0 and detected > 0 in all cells
rowdata_df_filtered <- rowdata_df_filtered %>% 
  dplyr::filter(gene_id %in% common_genes) %>%
  dplyr::filter(mean > 0 & detected > 0)
```


```{r}
# create upset plot 
gene_detect_df <- rowdata_df_filtered %>%
  dplyr::group_by(gene_id) %>%
  dplyr::summarise(tools_detected = list(unique(tool)))

ggplot(gene_detect_df, aes(x = tools_detected)) +
  geom_bar() +
  scale_x_upset(n_intersections = 3)
```

Similar to with the single-cell data most genes are detected in both Alevin-fry and Spaceranger which is good. 

## Concluding thoughts 

- Alevin-fry and Spaceranger only result in similar distributions of UMI/spot and genes/spot, although it is not quite as nice as the overlay you see with single-cell libraries. 
- Both tools also show high correlation in mean gene expression and high overlap in the genes that are detected.

## Session Info

```{r}
sessioninfo::session_info()
```

