---
title: "Compare Fasta and GTF Files"
output: html_notebook
---

## Setup

### Load Libraries for import

```{r setup}
library(Biostrings)
library(BSgenome)
library(eisaR)
library(GenomicFeatures)
library(tidyverse)
library(rtracklayer)
```

### File and Directory Setup

```{r}
## get file directories
ref_dir <- file.path("..", "..", "workflows", "rnaseq-ref-index")

txome_fasta <- file.path(ref_dir, "fasta", "Homo_sapiens.GRCh38.103.spliced.txome.fa.gz")
ensembl_fasta <- file.path(ref_dir, "fasta", "Homo_sapiens.GRCh38.txome.fa.gz")
ncrna_fasta <- file.path(ref_dir, "fasta", "Homo_sapiens.GRCh38.ncrna.fa.gz")
```

### Load fastas and gtf files for comparisons

```{r}
# read in fastas
txome_genome <- Biostrings::readDNAStringSet(txome_fasta)
names(txome_genome) <- sapply(strsplit(names(txome_genome), " "), .subset, 1)

ensembl_genome <- Biostrings::readDNAStringSet(ensembl_fasta)
names(ensembl_genome) <- sapply(strsplit(names(ensembl_genome), " "), .subset, 1)

ncrna_genome <- Biostrings::readDNAStringSet(ncrna_fasta)
names(ncrna_genome) <- sapply(strsplit(names(ncrna_genome), " "), .subset, 1)
```

```{r}
# load in the gtf files
txome_gtf_file <- file.path(ref_dir, "annotation", "Homo_sapiens.GRCh38.103.spliced.txome.gtf")
genome_gtf_file <- file.path(ref_dir, "annotation", "Homo_sapiens.GRCh38.103.gtf")

txome_gtf <- rtracklayer::import(txome_gtf_file)
genome_gtf <- rtracklayer::import(genome_gtf_file)
```

## Fasta Comparisons

Start by looking at the fasta files and which sequences are shared across 
the fasta directly from ensembl vs. the fasta created from 
eisaR::getFeatureRegions for spliced regions only. 

```{r echo = FALSE}
message("Ensembl fasta w/ncRNA") 
length(ensembl_genome)
head(ensembl_genome)

message("Spliced gene only from eisaR") 
length(txome_genome)
head(txome_genome)
```

One concern is that any differences could be from loss of ncRNA, so also going 
to take a look at those for comparison. 
```{r}
length(ncrna_genome)
head(ncrna_genome)
```

First, I wanted to check and see how many of the sequences are overlapping and 
what sequences we might be losing. 

```{r}
## remove the version numbers from ensembl_genome
names(ensembl_genome) = gsub("\\..*", "", names(ensembl_genome))
```


```{r}
## how many transcripts are shared
shared_tx= intersect(names(ensembl_genome), names(txome_genome))
length(shared_tx)
lost_tx <- ensembl_genome[!names(ensembl_genome) %in% shared_tx]

length(lost_tx)
head(lost_tx)
```

At first glance it looks like the missing transcripts are really short, so maybe
it doesn't matter if we lose them. Let's look at the length of these transcripts
in comparison to the transcripts we do capture and see if they are in fact much
shorter. 

```{r}
ensembl_width_df <- data.frame("width" = width(ensembl_genome), 
                               "group" = "Ensembl")
txome_width_df <- data.frame("width" = width(txome_genome),
                             "group" = "EisaR_Txome")
lost_width_df <- data.frame("width" = width(lost_tx),
                            "group" = "Lost Tx")

all_width_df <- bind_rows(ensembl_width_df,
                          txome_width_df, lost_width_df)
```

```{r}
ggplot(all_width_df, aes(x = width, color = group)) +
  geom_density() + 
  theme_classic() + 
  xlim(c(0,4000))
```
From this plot you can see that there are still a lot of transcripts that are 
of similar size to the main transcriptome that are lost, so it isn't just small
transcripts that are getting lost. 


How many of the missing transcripts are ncRNA transcripts? 
```{r echo = FALSE}
# how many ncrna sequences are present in the new transcriptome? 
names(ncrna_genome) = gsub("\\..*", "", names(ncrna_genome))
shared_ncrna= intersect(names(ncrna_genome), names(txome_genome))

message("Ensembl ncRNA fasta") 
length(ncrna_genome)

message("ncRNA found in both eisaR fasta and ncRNA.fa from ensembl")
length(shared_ncrna)

```

```{r}
## how many of the lost transcripts are ncRNA transcripts? 
length(intersect(names(lost_tx), names(ncrna_genome)))

```

It looks like only 1/4 of the missing transcript sequences are from the ncRNA.fa
file. So that doesn't account for the other 75% of the sequences. Can we figure
out what type of sequences are missing? 

## GTF Comparisons

I then looked at the gtf files to see what type of transcript sequences 
are missing. 

```{r}
## subset gtf by transcripts in the fasta file made from combining cdna/ncrna.fa
genome_gtf <- genome_gtf[genome_gtf$transcript_id %in% 
                           names(ensembl_genome)]

# find the features that are in the eisaR gtf and also in the ensembl gtf
shared_gtf <- genome_gtf[genome_gtf %in% txome_gtf]

# get the features only found in the eisaR txome
txome_only_gtf <- txome_gtf[!txome_gtf %in% genome_gtf]

# get the features only found in the ensembl gtf
genome_only_gtf <- genome_gtf[!genome_gtf %in% txome_gtf]

length(genome_gtf)
length(txome_gtf)
length(shared_gtf)
length(txome_only_gtf)
length(genome_only_gtf)
```

It looks like if we were to use the ensembl version, we would be missing around
500,000 annotations. Additionally, the eisaR version includes ~17,000 
annotations that are not found in the ensembl gtf. 

So what type of annotations would we be missing if we were to use the eisaR 
version over the ensembl version?

```{r}
# look specifically at features in the ensembl gtf that are not in the eisaR gtf
table(genome_only_gtf$gene_biotype)
```

It looks like we are missing almost 500,000 protein coding regions, which seems
relatively important.

Are these protein coding genes corresponding to loss of genes from or loss of 
transcripts? 

```{r}
# genes in ensembl gtf
ensembl_genes <- unique(genome_gtf$gene_id)
length(ensembl_genes)

# genes in eisaR gtf
txome_genes <- unique(txome_gtf$gene_id)
length(txome_genes)

# how many are found in both
shared_genes <- intersect(ensembl_genes, txome_genes)
length(shared_genes)

# how many are found in ensembl that are missing from eisaR? 
lost_genes <- genome_gtf[!genome_gtf$gene_id %in% shared_genes]
length(lost_genes)

# how many are found in eisaR that are missing from ensembl?
txome_only_genes <- txome_gtf[!txome_gtf$gene_id %in% shared_genes]
length(txome_only_genes)
```

It looks like the eisaR created gtf has all of the genes that are found in 
ensembl and then some. What are the new genes that are added? 

```{r}
txome_only_genes$gene_id[1:20]
```

Many of these genes show up as uncategorized. 

Based on this brief comparison, it looks like there are quite a lot of 
differences between the transcriptome.fa obtained directly from ensembl and the 
transcriptome.fa obtained by subsetting the ensembl gtf for only regions 
corresponding to spliced mRNA. Also in looking through a few of the genes, 
there doesn't appear to be any clear type of transcripts that are lost. One 
hypothesis was that they would belong to genes with one exon, but many are 
transcripts from genes with multiple exons (based on searching through ensembl).
It also appears that the standard is to use the fasta and gtf directly from 
ensembl in scRNA-seq. Based on the information found here, it doesn't seem to 
be in our best interest to change what we are using to generate our reference 
files for scRNA-seq. One potential problem that might come up is if the 
reference files being generated this way for snRNA-seq will be adequate in 
comparison to using the full genome with cellranger. There is no clear gold 
standard for snRNA-seq index generation with Alevin or Kallisto, but the 
approach of combining cDNA sequences with flanking intronic sequences was the 
recommended approach for both. 

## Session info

```{r paged.print=FALSE}
sessioninfo::session_info()
```
